pipert/core/message.py

```diff
@@ -0,0 +1,107 @@
+from abc import ABC, abstractmethod
+from io import BytesIO
+
+import numpy as np
+import time
+import pickle
+import cv2
+
+from PIL import Image
+
+
+class Payload(ABC):
+
+    def __init__(self, data):
+        self.data = data
+        pass
+
+    @abstractmethod
+    def decode(self):
+        pass
+
+    @abstractmethod
+    def encode(self):
+        pass
+
+
+class FramePayload(Payload):
+
+    def __init__(self, data):
+        super().__init__(data)
+
+    def decode(self):
+        self.data = np.array(Image.open(BytesIO(self.data)))
+
+    def encode(self):
+        _, buf = cv2.imencode(".jpg", self.data)
+        self.data = buf.tobytes()
+
+
+class PredictionPayload(Payload):
+    def __init__(self, data):
+        super().__init__(data)
+
+    def decode(self):
+        pass
+
+    def encode(self):
+        pass
+
+
+class Message:
+    next_msg_id = 0
+
+    def __init__(self, data, stream_address):
+        if isinstance(data, np.ndarray):
+            self.payload = FramePayload(data)
+        else:
+            self.payload = PredictionPayload(data)
+        self.stream_address = stream_address
+        self.history = {}
+        self.id = Message.next_msg_id
+        Message.next_msg_id += 1
+
+    def update_payload(self, data):
+        self.payload.data = data
+
+    def get_payload(self):
+        return self.payload.data
```

Comment:
When you access the payload, you don't know if it's encoded or not. Maybe add a flag to the payload and on the first call to `get_payload` it should encode the payload automatically, unless explicitly told not to?

Reply:
I don't think this is necessary, because the `get_payload` function is not accessible when the payload is encoded, because when we encode the payload, we also encode the whole message object immediately afterwards. @ItamarWilf

Reply:
Hmm, your'e right, but I think it is still a nice feature to have. Maybe we don't want to spend CPU time on decoding a payload that ends up not being used because of non-blocking queues or something like that. What do you think @sionovd ? Maybe this isn't required right now but if we don't solve it now we should open a new issue for it.